packages/ra-no-code/src/ResourceBuilderContext/useResources.ts

```diff
@@ -0,0 +1,131 @@
+import { useEffect, useMemo, useState } from 'react';
+
+type ResourceDefinition = {
+    name: string;
+    label?: string;
+};
+
+type ResourceDefinitionState = ResourceDefinition[] | undefined;
+
+type ResourceDefinitionStateActions = {
+    addResource: (resourceDefinition: ResourceDefinition) => void;
+    updateResource: (
+        name: string,
+        resourceDefinition: Partial<Omit<ResourceDefinition, 'name'>>
+    ) => void;
+    removeResource: (resourceDefinition: ResourceDefinition) => void;
+    setResources: (
+        value: (prevState: ResourceDefinitionState) => ResourceDefinitionState
+    ) => void;
+};
+const storageKey = '@@ra-no-code.resources';
+
+export const useResources = (): [
+    ResourceDefinitionState,
+    ResourceDefinitionStateActions
+] => {
+    const [resources, setInternalResources] = useState<ResourceDefinition[]>();
+
+    useEffect(() => {
+        const storedResourceDefinitions = window.localStorage.getItem(
+            storageKey
+        );
+
+        if (!storedResourceDefinitions) {
+            return;
+        }
+
+        const resourceDefinitions = JSON.parse(storedResourceDefinitions);
+        setInternalResources(resourceDefinitions);
+    }, []);
+
+    const actions = useMemo(() => {
+        const setResources = (
+            value: (
+                prevState: ResourceDefinitionState
+            ) => ResourceDefinitionState
+        ) => {
+            setInternalResources(prevState => {
+                const newState = value(prevState);
+
+                if (newState != undefined) {
+                    window.localStorage.setItem(
+                        storageKey,
+                        JSON.stringify(newState)
+                    );
+                } else {
+                    window.localStorage.removeItem(storageKey);
+                }
+                return newState;
+            });
+        };
+        return {
```

Comment:
I think this file would be more readable if you defined the functions first, then returned the object containing them.